core/modules/replica/ServiceProvider.cc

```diff
@@ -28,27 +28,114 @@
 #include <stdexcept>
 
 // Qserv headers
+#include "lsst/log/Log.h"
 #include "replica/ChunkLocker.h"
 #include "replica/Configuration.h"
-#include "replica/DatabaseServices.h"
+#include "replica/DatabaseServicesPool.h"
+#include "replica/Messenger.h"
 #include "replica/QservMgtServices.h"
 
+namespace {
+
+LOG_LOGGER _log = LOG_GET("lsst.qserv.replica.ServiceProvider");
+
+} /// namespace
+
 namespace lsst {
 namespace qserv {
 namespace replica {
 
 ServiceProvider::Ptr ServiceProvider::create(std::string const& configUrl) {
+
     auto ptr = ServiceProvider::Ptr(new ServiceProvider(configUrl));
+
     // This initialization is made a posteriori because the shared pointer
     // onto the object can't be accessed via the usual call to shared_from_this()
     // inside the contsructor.
+
     ptr->_qservMgtServices = QservMgtServices::create(ptr);
+    ptr->_messenger        = Messenger::create(ptr, ptr->_io_service);
+
     return ptr;
 }
 
 ServiceProvider::ServiceProvider(std::string const& configUrl) {
     _configuration    = Configuration::load(configUrl);
-    _databaseServices = DatabaseServices::create(_configuration);
+    _databaseServices = DatabaseServicesPool::create(_configuration);
+}
+
+void ServiceProvider::run() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "run");
+
+    util::Lock lock(_mtx, context() + "run");
+
+    // Check if the service is still not running
+
+    if (_threads.size() != 0) return;
+
+    // Initialize BOOST ASIO servces
+
+    _work.reset(new boost::asio::io_service::work(_io_service));
+
+    auto self = shared_from_this();
+
+    _threads.clear();
+    for (size_t i = 0; i < config()->controllerThreads(); ++i) {
+        _threads.emplace_back(std::make_shared<std::thread>(
+            [self] () {
+
+                // This will prevent the I/O service from exiting the .run()
+                // method event when it will run out of any requests to process.
+                // Unless the service will be explicitly stopped.
+                self->_io_service.run();
+            }
+        ));
+    }
+}
+
+bool ServiceProvider::isRunning() const {
+
+    util::Lock lock(_mtx, context() + "isRunning");
```

Comment:
This should be more efficient if you don't build new strings on every call here. Same for other uses of `Lock`.

Reply:
Method ::context() is called just a few times in the lifetime of an application - first time from method ::run() when the application is being initialized, then from method ::stop()  (when the application is finishing in a clean way, not by being killed). There may be a couple calls to ::context() when method :;isRunning() is called from either of the above mentioned methods. 
So, in general adding two strings may be inefficient, in practice - this makes no difference on the performance of the application. Perhaps just a few extra microseconds.